types/plugin.ts

```diff
@@ -37,3 +43,302 @@ export const Plugins: Record<PluginID, Plugin> = {
 };
 
 export const PluginList = Object.values(Plugins);
+
+export interface PluginApiOperation {
+  operationId: string;
+  nameForModel: string;
+  serverUrl: string;
+  apiPath: string;
+  method: string;
+  summary?: string;
+  parameters?: [
+    {
+      name: string;
+      in: string;
+      description?: string;
+      required?: boolean;
+      schema: {
+        type: string;
+      };
+    },
+  ];
+  requestBody?: {
+    required?: boolean;
+    content: {
+      [key: string]: {
+        schema: {
+          type: string;
+          required?: string[];
+          properties: {
+            [key: string]: {
+              type: string;
+              description?: string;
+              required?: boolean;
+            };
+          };
+        };
+      };
+    };
+  };
+  responses?: {
+    [key: string]: {
+      description?: string;
+      content?: {
+        [key: string]: {
+          schema: {
+            type: string;
+            required?: string[];
+            properties?: {
+              [key: string]: {
+                type: string;
+                description?: string;
+              };
+            };
+          };
+        };
+      };
+    };
+  };
+}
+
+export interface PluginApiOperationList {
+  [operationId: string]: PluginApiOperation;
+}
+
+export function resolveRef(ref: string, document: any): any {
+  if (!ref.startsWith('#')) {
+    throw new Error(`Unsupported reference: ${ref}`);
+  }
+
+  const refPath = ref.substring(1).split('/');
+  let current = document;
+  for (const refPathPart of refPath) {
+    if (refPathPart === '') {
+      continue;
+    }
+    if (current === undefined) {
+      throw new Error(`Cannot resolve reference: ${ref}`);
+    }
+    current = current[refPathPart];
+  }
+  return current;
+}
+
+export async function getPluginApiOperationsFromUrl(
+  url: string,
+): Promise<PluginApiOperationList | null> {
+  const pluginUrl = getPluginUrl(url);
+  const plugin = await fetch(pluginUrl, { method: 'GET' }).then((response) =>
+    response.json(),
+  );
+  const result = await fetch(plugin.api.url, { method: 'GET' });
+  const yamlContent = await result.text();
+
+  const yaml = require('js-yaml');
+
+  const document = yaml.load(yamlContent);
+  if (!document) return null;
+
+  const pluginApiOperationList: PluginApiOperationList = {};
+
+  for (const path in document.paths) {
+    const apiPath = path;
+    for (const method in document.paths[path]) {
+      const operationObject = document.paths[path][method];
+      let serverUrl;
+      if (document.servers && document.servers[0]?.url) {
+        serverUrl = document.servers[0]?.url;
+      } else {
+        serverUrl = url.substring(0, url.lastIndexOf('.well-known/ai-plugin.json'));
```

Comment:
This `substring()` parsing logic seems to fail with some plugins. For example, "AskYourPDF" (`https://plugin.askyourpdf.com`).
I would suggest using the following instead:
```ts
serverUrl = new URL(url).origin;
```